doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,562 @@
+a real move of the underlying structure can be performed or not. This could be
+the case for instance with two kinds of heap allocators or when two
+RangeAllocators are using the same range. If this is not the case the move
+operations will be replaced with an expensive copy operation (fake move).
+
+Furthermore, the suggested zero copy types will only once acquire memory from an
+allocator. The reason is that we would like to guarantee zero copy throughout the
+usage of the container otherwise some intransparent copies may occur when the
+container allocates memory multiple times and requires one contiguous piece
+of memory which the allocator may not provide. (See behavior of `realloc` in C).
+
+### Implementation of a C++20 Concept in C++14
+The type trait can be implemented via the C++ SFINAE (substition failure is not
+an error) principle.
+```cpp
+template<typename T>
+struct IsAllocatorTypeTrait {
+    template <typename C, class = void>
+    struct HasAllocateMethod : std::false_type {};
+
+    template <typename C>
+    struct HasAllocateMethod<
+        C, std::void_t<std::enable_if_t<
+               std::is_same<decltype(std::declval<C>().allocate(
+                                std::declval<uint64_t>())),
+                            void>::value,
+               void>>> : std::true_type {};
+
+    // test all other methods
+    static_assert(HasAllocateMethod<T>::value, 
+              "Type is not an allocator, void* allocate(uint64_t) missing");
+};
+```
+
+## Design of Zero Copy Containers
+
+The design pursues the following goals.
+
+1. The `cxx::vector<T, Capacity>` should not change in its API.
+
+2. All static cxx containers are still owner of all their data. This means the
+   memory which the stack allocator manages must be stored inside the cxx
+   container class.
+
+3. A function developer should not have to specify the allocator type of a
+   container. This means the allocator type is independent of the container type.
+   This has to work:
+   ```cpp
+   void myFunkyFunction(const cxx::vector<int> & bla) {}
+
+   cxx::vector<int, 20> a;
+   cxx::vector<int, 30> b;
+   cxx::vector<int> c(Allocator::heap);
+
+   myFunkyFunction(a);
+   myFunkyFunction(b);
+   myFunkyFunction(c);
+   ```
+
+4. All containers must be shared memory compatible, this means no vtables and
+   virtual. Furthermore the usage of function pointers, `cxx::function` and
+   `cxx::function_ref` is forbidden in all containers and their underlying 
+   constructs.
+
+5. Before using any cxx container the user as to call `reserve()` once otherwise
+   the container will have a default capacity of zero. It is not allowed to
+   call `reserve()` multiple times since this could violate the zero copy
+   guarantee.
+   Stack based containers do not require a `reserve()` call beforehand but
+   should provide this call as well for compatibility reasons.
+
+   * It is not possible to provide solely the capacity (`reserve` parameter) via
```

Comment:
I think for this it is also important what kind of template types we support for say `vector`.